src/sap.ui.core/src/sap/ui/model/FilterProcessor.js

```diff
@@ -131,13 +131,8 @@ sap.ui.define(['jquery.sap.global'],
 						bLocalMatch = true;
 					}
 				}
-				if (bLocalMatch && oMultiFilter.bAnd) {
-					bMatched = true;
-				} else if (!bLocalMatch && oMultiFilter.bAnd) {
-					bMatched = false;
-					return false;
-				} else if (bLocalMatch) {
-					bMatched = true;
+				if ((!bLocalMatch && oMultiFilter.bAnd) || (bLocalMatch && !oMultiFilter.bAnd)) {
```

Comment:
Interestingly, I prepared nearly the same change internally but used the shorter condition
```
  if ( bLocalMatch != bMatched ) {
```
I also added a few more comments. 

If you don't mind, I'll take your commit as base and merge my condition and comments. I'll also add some unit test for this special case.